### DISCUSSION TITLE
Template sensors with multiple batteries

### DISCUSSION THREAD

**Discussion by jontraceyuk** [WillCodeForCats/solaredge-modbus-multi#646] [General]
Quick YAML question: taking the example code provided for the sensors, if I have more than one battery (in this case 3), do I need to add up the sum of the batteries

eg - the example code
- name: "Power - Solar Generation"
        unit_of_measurement: "W"
        device_class: "power"
        state_class: "measurement"
        # The net of DC power from inverter and battery, only if this is positive
        state: "{{ max([float(states('sensor.solaredge_i1_dc_power'), 0) + float(states('sensor.solaredge_b1_dc_power'), 0), 0]) }}"
        availability: "{{ has_value('sensor.solaredge_i1_dc_power') and has_value('sensor.solaredge_b1_dc_power') }}"

**Should i replace the line** state: "{{ max([float(states('sensor.solaredge_i1_dc_power'), 0) + float(states('sensor.solaredge_b1_dc_power'), 0), 0]) }}"
        availability: "{{ has_value('sensor.solaredge_i1_dc_power') and has_value('sensor.solaredge_b1_dc_power') }}"

**with something like** 

state: "{{ max([float(states('sensor.solaredge_i1_dc_power'), 0) + float(states('sensor.solaredge_b1_dc_power')**+ float(states('sensor.solaredge_b2_dc_power')+ float(states('sensor.solaredge_b3_dc_power')**,, 0), 0]) }}"
        availability: "{{ has_value('sensor.solaredge_i1_dc_power') and has_value('sensor.solaredge_b1_dc_power') }}"

**Comment by deqw** [comment]
Are these B2 and B3 sensors available anyway? I have two batteries which are in parallel on one (1) inverter and represent them as one (B1) battery.

**Reply by jontraceyuk** [reply]
They show as three entities, and have three charge and discharge rates but the system they operate as one battery system. For example if i discharge 6kw I can see 2kw from Battery 1 2 from battery 2 etc

**Reply by deqw** [reply]
Trying to get your installation clear. So these three batteries are plus/minus wired from 1 to 2, then to 3 and finally to your (one) inverter? And the data cable is daisy chained too?

**Reply by jontraceyuk** [reply]
Yes three physical batteries wires in serial form and presented to the inverter by a single pair of MC4 Connectors. From solar edge perspective its just one 29kw battery. But from home assistants point of view I have three batteries, each has a charge level, each has a charge and discharge rate. So when setting up templates do I need to add them together or do I just use Battery 1 and ignore 2 and 3. Hope that helps.

**Reply by deqw** [reply]
For your 'Quick YAML question': yes then you should add them up. But I still can't get a grip on your setup. Are these SolarEdge batteries? Instead of messing with the YAML I would look how to get your batteries presented as just one battery, which it in the end is.

**Reply by jontraceyuk** [reply]
Yes, all solaredge batteries (photo attached) all connected by one set of MC4 - the ones on the right in the inverter picture but as you can see in the third pic - present are three batteries in the home assistant
![IMG_4690](https://github.com/user-attachments/assets/52ea49d6-9ac9-4577-8a07-abce8f70567f)
![IMG_4691](https://github.com/user-attachments/assets/12369d17-3451-49ad-85b9-3494cc195749)
![Screenshot 2024-08-05 at 12 12 41](https://github.com/user-attachments/assets/1b40583c-a834-416b-98c6-b276123ef11c)

**Reply by jontraceyuk** [reply]
will try as you suggest and add up the battery outputs to get a working total

**Reply by cyberkryten** [reply]
Here's what I'm using.  Setting each Battery value into a variable helps debugging as you can just output {{myB1}} after pasting the code into the Developer Tools Template to check any values and it isn't really any less efficient (certainly not to the point it matters)  

However, be aware that these calculated totals will fractionally lag the actual values (because they must trigger and recalculate).  This can cause issues where you are using the total value when the charge/discharge starts/stops because the three batteries will have differing values.  So if you are doing calculations (eg to figure out the current Solar Generation) then I'd suggest using B1/B2/B3 values in that template instead of these calculated totals which may be slightly out of date as this will cause phantom spikes due to the lag.

This delay is made far worse if the batteries are connected to the inverter using CAT6 cables for the comms as it is far slower than the wireless SolarEdge Network in reacting to commands and reporting charge/discharge state.  It is bad enough that the inverter (in my case, at least) fails to sensibly control them when facing an export limit.  

I found this out the hard way and if you are using them cabled would recommend changing to wireless (which usually is a bad idea, but the speed of processing is just too slow on the physically cabled connection when multiple batteries are daisy chained together).

Template entires:

`
- sensor:
    - name: "Calc Battery All Charge"
      unique_id: calc_battery_all_charge
      unit_of_measurement: "kWh"
      device_class: energy
      state_class: total_increasing
      state: >
        {% set myB1 = float(states('sensor.solaredgemodbus_b1_energy_import'),0) %}
        {% set myB2 = float(states('sensor.solaredgemodbus_b2_energy_import'),0) %}
        {% set myB3 = float(states('sensor.solaredgemodbus_b3_energy_import'),0) %}
        {% set myvalue = myB1 + myB2 + myB3 %}
        {{ myvalue }}

- sensor:
    - name: "Calc Batteries All Charge Today"
      unique_id: calc_batteries_all_charge_today
      unit_of_measurement: "kWh"
      device_class: energy
      state_class: total_increasing
      state: >
        {% set myB1 = float(states('sensor.calc_solaredgemodbus_b1_import'),0) %}
        {% set myB2 = float(states('sensor.calc_solaredgemodbus_b2_import'),0) %}
        {% set myB3 = float(states('sensor.calc_solaredgemodbus_b3_import'),0) %}
        {% set myvalue = myB1 + myB2 + myB3 %}
        {{ myvalue }}

- sensor:
    - name: "Calc Battery All Discharge"
      unique_id: calc_battery_all_discharge
      unit_of_measurement: "kWh"
      device_class: energy
      state_class: total_increasing
      state: >
        {% set myB1 = float(states('sensor.solaredgemodbus_b1_energy_export'),0) %}
        {% set myB2 = float(states('sensor.solaredgemodbus_b2_energy_export'),0) %}
        {% set myB3 = float(states('sensor.solaredgemodbus_b3_energy_export'),0) %}
        {% set myvalue = myB1 + myB2 + myB3 %}
        {{ myvalue }}

- sensor:
    - name: "Calc Batteries All Discharge_today"
      unique_id: calc_batteries_all_discharge_today
      unit_of_measurement: "kWh"
      device_class: energy
      state_class: total_increasing
      state: >
        {% set myB1 = float(states('sensor.calc_solaredgemodbus_b1_export'),0) %}
        {% set myB2 = float(states('sensor.calc_solaredgemodbus_b2_export'),0) %}
        {% set myB3 = float(states('sensor.calc_solaredgemodbus_b3_export'),0) %}
        {% set myvalue = myB1 + myB2 + myB3 %}
        {{ myvalue }}

- sensor:
    - name: "Calc Battery All State"
      unique_id: calc_battery_all_state
      unit_of_measurement: "%"
      state: >
        {% set myB1 = float(states('sensor.solaredgemodbus_b1_state_of_energy'),0) %}
        {% set myB2 = float(states('sensor.solaredgemodbus_b2_state_of_energy'),0) %}
        {% set myB3 = float(states('sensor.solaredgemodbus_b3_state_of_energy'),0) %}
        {% set myvalue = ((myB1 + myB2 + myB3) / 3) | round(0) %}
        {{ myvalue }}

# Calculate Total Battery Power Values
- sensor:
    - name: "SEModbus Power - Batteries DC Power"
      unique_id: semodbus_power_batteries_dc_power
      unit_of_measurement: "W"
      device_class: "power"
      state_class: "measurement"
      state: >
        {% set myB1 = float(states('sensor.solaredgemodbus_b1_dc_power'),0) %}
        {% set myB2 = float(states('sensor.solaredgemodbus_b2_dc_power'),0) %}
        {% set myB3 = float(states('sensor.solaredgemodbus_b3_dc_power'),0) %}
        {% set myvalue = ((myB1 + myB2 + myB3)) %}
        {{ myvalue }}

# Total Energy Stored in the Batteries
- sensor:
    - name: "Calc Battery Total Capacity"
      unique_id: calc_battery_total_capacity
      unit_of_measurement: kWh
      state: >
        {% set myB1 = float(states('sensor.solaredgemodbus_b1_maximum_energy'),0) %}
        {% set myB2 = float(states('sensor.solaredgemodbus_b2_maximum_energy'),0) %}
        {% set myB3 = float(states('sensor.solaredgemodbus_b3_maximum_energy'),0) %}
        {% set myvalue = ((myB1 + myB2 + myB3)) %}
        {{ myvalue }}

# Current Energy Stored in the Batteries
- sensor:
    - name: "Calc Battery Current Capacity"
      unique_id: calc_battery_current_capacity
      unit_of_measurement: kWh
      state: >
        {% set myCurrentCapacity = (float(states('sensor.calc_battery_all_state'),0) / 100) * float(states('sensor.calc_battery_total_capacity'),0) %}
        {{ myCurrentCapacity }}
`

**Reply by cyberkryten** [reply]
Ah, so marking as code doesn't turn off markdown.  Bullet points are a hyphen, Bold titles were comments which started with a hash

**Reply by deqw** [reply]
Thanks for the pictures. Got it now. Completely different from the 48V/10kWh setup I have here. Was just writing an example for you to make your yaml more readable, but @cyberkryten is (way) ahead of me. Thanks for posting the code!

**Reply by jontraceyuk** [reply]
Thanks both will have a play
